mcm/json_layer/validation.py

```diff
@@ -0,0 +1,574 @@
+            for request in requests
+        }
+        threads_to_run = set()
+        for config in thread_config_requests.values():
+            scenarios = config["scenarios"]
+            threads_for_request = {int(thread_str) for thread_str in scenarios.keys()}
+            threads_to_run.update(threads_for_request)
+
+        if threads_to_run:
+            # Discard the minimum as it was already sent
+            threads_to_run.discard(min(threads_to_run))
+            for threads_int in threads_to_run:
+                validation_control.submit_item(validation_name, threads_int)
+
+    def get_event_count_for_validation(self, request, with_explanation=False, threads=None) -> int | tuple[int, str]:
+        threads_int = threads if isinstance(threads, int) else 1
+        threads_str = str(threads_int)
+        # Run the validation job based on output events
+        test_on_output = self._test_on_output_events()
+        # Validation scenarios
+        validation_thread_configurations = self._get_thread_configurations(request=request)
+        validation_scenarios = validation_thread_configurations["scenarios"]
+        thread_config = validation_scenarios.get(threads_str)
+        if not thread_config:
+            self.logger.error("Unable to retrieve the validation scenario for %s threads", threads_str)
+            self.logger.warning("Setting a default number of events to be able to generate a script to manually test it!")
+            if with_explanation:
+                not_found_message = f"# Short validation strategy: Unable to retrieve the validation scenario for {threads_str} threads"
+                return 1, not_found_message
+            else:
+                return 1
+
+        target_input_events = thread_config["target_input_events"]
+        target_output_events = thread_config["target_output_events"]
+        explanation = thread_config["explanation"]
+        events_to_return = target_output_events if test_on_output else target_input_events
+        if not with_explanation:
+            return events_to_return
+
+        if test_on_output:
+            explanation.append("# This validation will be computed based on the target output events!")
+        else:
+            explanation.append("# This validation will be computed based on the target input events!")
+
+        return events_to_return, '\n'.join(explanation)
+
+    def tweak_cmsdriver_for_validation(self, request, sequence_dict, threads) -> dict:
+        threads_int = threads if isinstance(threads, int) else 1
+        events = self.get_event_count_for_validation(request=request, threads=threads)
+        test_on_output = self._test_on_output_events()
+
+        # Set the number of event parameters
+        sequence_dict["nThreads"] = threads_int
+        if test_on_output:
+            sequence_dict["number_out"] = events
+            sequence_dict["number"] = events
```

Comment:
these two numbers cannot have the same values. this is causing indeed hard truncation of validation jobs.
it should be set to -1 to let events be generated until "events" are found in output

Reply:
-1 does not function with wmlhe based requests because of the external producer. but the two values cannot be the same, in particular and number_out is meant to be 100, and number ten or hundred times bigger. see suggestions above